### PR TITLE
Add API fetch error handling

### LINKED ISSUE #1180: Handle API errors for blocks that make requests (All Products, Reviews...)
In #1166 we added a friendly error message when a JS-rendered block throws a JS error, but we would like to show this same message (or a similar one) when an API request made by one of this blocks returns an error.

### PR DESCRIPTION
**Pull request by mikejolley**
This is my attempt to fix #1180 based on the experiment @nerrad started in #1248.

I'm not sure if there is more to do here, but essentially the `useCollection` hook will throw an exception if it gets an `error` response back from the API request (once resolved). This is picked up by `BlockErrorBoundry`. I have updated `BlockErrorBoundry` to show errors by default but I can revert this is not desirable (see screenshot below).

I did have issues getting this to work at first because @nerrad and I were trying to throw exceptions from within hooks. Apparently this doesn't work (I got exceptions, but not the one I was throwing!). I found a workaround in this issue: https://github.com/facebook/react/issues/14981 using `setState`.

Fixes #1180

### Screenshots

This is an example error message when the API returns an error:

Screenshot 2020-01-30 at 13 13 50

### How to test the changes in this Pull Request:

Cause an API error; I hardcoded this into the products API `get_items()` method to test:

```
return new RestError( 'woocommerce_rest_product_invalid_id', __( 'Invalid product ID.', 'woo-gutenberg-products-block' ), array( 'status' => 404 ) );
```

Then view the all products block.

### Changelog

> Added error handling for API calls.

### CODE REVIEW

**Comment by mikejolley** [comment]
Looks like I need to update tests but I think a review first would be best so I'm not changing tests based on things that need revisions.

File: assets/js/base/components/block-error-boundary/block-error.js [hunk 1]
```diff
@@ -40,7 +40,7 @@ BlockError.propTypes = {
 	/**
 	 * Error message to display below the content.
 	 */
-	errorMessage: PropTypes.string,
+	errorMessage: PropTypes.oneOfType( [ PropTypes.string, PropTypes.node ] ),
```
Comment (Aljullu) [inline assets/js/base/components/block-error-boundary/block-error.js]: `PropTypes.node` already includes `PropTypes.string`, so no need to include both.
```suggestion
	errorMessage: PropTypes.node,
```

File: assets/js/base/components/block-error-boundary/style.scss [hunk 2]
```diff
@@ -14,17 +14,23 @@
...
+}
 
 @include breakpoint( ">480px" ) {
 	.wc-block-error {
 		flex-direction: row;
 	}
 
 	.wc-block-error__image + .wc-block-error__content {
-		margin-left: $gap;
+		margin-left: $gap-larger;
```
Comment (Aljullu) [inline assets/js/base/components/block-error-boundary/style.scss]: Personally speaking here, but I feel this makes the space too wide. If the previous one was too narrow, what do you think about a middle point (`$gap-large`)?
```suggestion
		margin-left: $gap-large;
```

File: assets/js/base/hooks/use-collection.js [hunk 3]
```diff
@@ -70,6 +74,23 @@ export const useCollection = ( options ) => {
...
+			const error = store.getCollectionError( ...args );
+			if ( error ) {
+				// Throw an exception within setState - this is needed because
+				// this is a hook (https://github.com/facebook/react/issues/14981).
+				setState( () => {
+					throw {
+						message: (
+							<Fragment>
+								<strong>{ error.status }</strong>:&nbsp;
+								{ error.statusText }
```
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/hooks/use-collection.js]: Moving the error handling to the boundary works well 👍

The other change however, not so well. I can't get it to pick up the `reject` from a try catch block. I think the approach might not be compatible with the `getCollection` generator method. That does the fetch, but I can't avoid setting state there - throwing exceptions from that method doesn't appear to work...

[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/hooks/use-collection.js]: Ok; I added reject back too. But we still need `receiveCollectionError`. That was what you wanted to get rid of to be clear?

File: assets/js/base/components/block-error-boundary/index.js [hunk 4]
```diff
@@ -14,7 +14,24 @@ class BlockErrorBoundary extends Component {
...
+					</Fragment>
+				),
+				hasError: true,
+			};
+		}
+
+		if ( typeof error.message !== 'undefined' ) {
+			return { errorMessage: error.message, hasError: true };
+		}
 	}
```
Comment (nerrad) [inline assets/js/base/components/block-error-boundary/index.js]: This returns undefined if none of the conditions match, should probably default to a more generic message as a fallback?

Comment (mikejolley) [reply assets/js/base/components/block-error-boundary/index.js]: The message is not required; we can return an undefined errorMessage though.

**Comment by mikejolley** [comment]
Tests are passing 🎆

**Pull request merged**